meshmode/array_context.py

```diff
@@ -437,77 +437,73 @@ def _get_array_container_ac_arith(ary: ArrayContainerWithArithmetic):
 
 # {{{ dataclass containers
 
-# FIXME: Document what this is good for
+# TODO Will this interfere with type-checking, should we want it later?
+# We could always change the definition based on typing.TYPE_CHECKING...
 class NumpyObjectArray(np.ndarray):
-    pass
-
-
-class DataclassArrayContainer(ArrayContainer):
-    """An :class:`ArrayContainer` that implements serialization and
-    deserialization of :func:`~dataclasses.dataclass` fields.
+    """Used as an attribute type with :func:`dataclass_array_container` to
+    document that the attribute will be an object array.
+    :func:`dataclass_array_container` will treat this attribute as an
+    array container.
     """
 
 
-class DataclassArrayContainerWithArithmetic(
-        ArrayContainerWithArithmetic,
-        DataclassArrayContainer):
-    """An :class:`DataclassArrayContainer` where each field is assumed to
-    support arithmetic, e.g. through :class:`ArrayContainerWithArithmetic`.
-    """
+_ARRAY_CONTAINER_LIKE = (ArrayContainer, NumpyObjectArray)
 
 
-@memoize(key=lambda ary: (type(ary), _serialize_dataclass_container))
-def _generate_dataclass_serialization_func(ary):
-    from pytools.codegen import remove_common_indentation
-    code = remove_common_indentation("""
-    def _generated_serialize_dataclass(_ary):
-        return [{}]
-    """.format(", ".join(
-        f"('{f.name}', _ary.{f.name})" for f in fields(ary)
-        if issubclass(f.type, (ArrayContainer, NumpyObjectArray))
-        ))
-    )
+def dataclass_array_container(cls):
+    """A class decorator that makes the class to which it is applied a
+    ``frozen`` :func:`~dataclasses.dataclass` and registers appropriate
+    implementations of :func:`serialize_container` and :func:`deserialize_container`.
 
-    exec_dict = {}
-    exec(compile(code, "<generated code>", "exec"), exec_dict)
-    exec_dict["_MODULE_SOURCE_CODE"] = code
+    Attributes that are not array containers are allowed. In order to decide
+    whether an attribute is an array container, the declared attribute type
+    is checked for whether it is a subclass of :class:`ArrayContainer`.
+    Note that this classification does not necessarily agree with
+    :class:`is_array_container` (which is not usable here because it operates
+    on instances, not types). :mod:`numpy` object arrays are a particularly
+    important example of this. To document that an attribute *is* a
+    :mod:`numpy` object array, use :class:`NumpyObjectArray` in the
+    attribute type annotation.
```

Comment:
This general bit of ugliness is also perhaps worth a look. Maybe we should make it so that array containers can be recognized by their type? But then we get all of `np.ndarray`... Maybe that's OK?

Reply:
Yeah, it seems to be heading that way. I'm mostly fine with allowing all of `ndarray` but erroring out when trying to serialize it?

Reply:
Actually not so much on the erroring out part, as it seems dishonest if we said it's an array container already.

Reply:
I've kept it at erroring out. We can continue this discussion post-merge.